Live Animation, Receive data

janga1997 opened the pull request:
I tried to use existing work done by Doug Baldwin, and making it work with a post-processor as much as I can (Doug's code already had data on nodes/elements/drivers/sockets before launching the live simulation, but we have to parse it from the log files.)

Here's how the workflow goes,

1. Select the file for MBDyn simulation just as before, using the `Select Input File` button.
2. Configure any extra options as you want before (env variables, cmd line options, output filename, etc)
3. Start the simulation by pressing `Run Simulation`.
-------------------------------------------------------------------------------------------------------------------------------------------
4. This is the part where I would like your review on.
The user has to make sure that when using the `stream output` element ( The `stream motion output` will be deprecated sometime, as stated in the manual ), 

- He/She always uses `create: no` . I guess we could check for that in Blendyn and make it work accordingly, but its another layer of complication, that could be simply averted.
- He/She specifies the order of the nodes to the stream the same as their definition. I guess this could be averted by **using `all` instead of manually specifying nodes**, but I haven't been able to make that keyword work. Maybe I'm doing something wrong.
--------------------------------------------------------------------------------------------------------------------------------------------

5. Load in the nodes and ***elements** by pressing `Load .log file` as before. I've placed the same button down too, to make it easier to work. then add the nodes/elements to the scene.
**EDIT** - The elements are logged in the log file in after atleast one message is sent to the file drivers through the sockets. (Since file drivers are defined before elements).
Since we are now using a dedicated file driver, you have to press `Load .log file` button again after the starting sockets in the next step

Add them to the scene by pressing `Add nodes to scene`.

6. Just press the `Start Sockets` button, to start the receiver and sender sockets.

7. Load in the elements now by pressing `Load .log file`.
Add them to the scene by pressing `Add all the elements to the scene.`

8. Press **"P"** (case doesn't matter) to start the simulation. You can press the same key to pause/continue the simulation at any point.
CAUTION - Keep in mind the timeouts you may have set. If your "pause" exceeds that duration, the simulation will stop.
Keyframes will be generated and stored in Blender scene for every step of the simulation. This way you can revisit previous points in the simulation, or render a video right away.

9. If you want to interrupt the process at any point, just press `ESC`.

You can repeat the process by pressing `Run Simulation` and `Start Sockets` again and again in succession.

janga1997 committed:
Live Animation, Receive data

janga1997 committed:
remove print

janga1997 commented:
If something goes wrong, start the MBDyn in a console, and start the sockets inside Blender too observe any possible errors in input .mbd files itself.

janga1997 committed:
Pause live sim and keyframes in-sync

janga1997 committed:
Edit sender and receiver port numbers from UI

janga1997 commented:
**Update**
I think its pointless to work anymore on this feature until we add File Drivers and Stream output elements to the log file. (The next obvious feature being , figuring out a way to send custom to data to File Drivers (which might serve as input to nodes/elements) through Blender)
That is the only way I can look for them, and generate appropriate sockets.
So, I will be working on that, and also adding the missing joint elements we talked about earlier.

louisgag commented:
Can you send me the input file, with the stream output element, that you 
use to test?

On 19-Jul-2017 01:44 PM, VSN Reddy Janga wrote:
> *Update*
> I think its pointless to work anymore on this feature until we add File 
> Drivers and Stream output elements to the log file. (The next obvious 
> feature being , figuring out a way to send custom to data to File 
> Drivers (which might serve as input to nodes/elements) through Blender)
> That is the only way I can look for them, and generate appropriate sockets.
> So, I will be working on that, and also adding the missing joint 
> elements we talked about earlier.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zanoni-mbdyn/blendyn/pull/30#issuecomment-316359646>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ALuOfeuTgs3-hHuGJFaX1m3OXIQYuRavks5sPewPgaJpZM4OZXEb>.
> 

-- 
Louis Gagnon, Ph.D.
Postdoctoral fellow
Dipartimento di Scienze e Tecnologie Aerospaziali
Politecnico di Milano
+39 02 2399 8642
http://louisgagnon.com/research/

janga1997 commented:
[double_rigid_pendulum_2.txt](https://github.com/zanoni-mbdyn/blendyn/files/1159084/double_rigid_pendulum_2.txt)
Here is a file for a double rigid pendulum.

louisgag commented:
I'm having a bit of difficulty running it. Does one have to run the simulation outside of Blender before running the sockets? Otherwise, the load log command gives an error. Also, it won't let me do anything for a while after having pressed start socket.

Here's my procedure:

- select input file (file alone, nothing ran)
- skipping 2 & 3
- start sockets
  - blender becomes unresponsive for 20 seconds
  - then this error appears:
```
timed out
Traceback (most recent call last):
  File "/home/louis/.config/blender/2.69/scripts/addons/blendyn-master/base.py", line 1313, in execute
    self.sender.send([1])
  File "/home/louis/.config/blender/2.69/scripts/addons/blendyn-master/liveanim.py", line 65, in send
    self.socket.send(pack('d'*len(floats), *floats))
AttributeError: 'NoneType' object has no attribute 'send'

location: <unknown location>:-1

location: <unknown location>:-1
parse_log_file(): Trying to read nodes and elements from file: /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2.log
Could not locate the file /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2.log.
Could not locate the file /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2.out
Could not locate the file /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2.rfm
Could not locate the .out file
Warning: Could not locate the .out file
```
- doing another iteration of start socket / load log file does not seem to help

janga1997 commented:
@louisgag I'm really sorry about the steps I laid down.
I forgot to add the crucial step of starting the simulation by pressing `Run Simulation`.
Edit1 - I  updated the steps I laid down.

Edit2 - I guess I should catch all those errors with appropriate report messages to prevent future confusion.

Every error you pointed out is because the simulation wasn't even started.

louisgag commented:
Ok. Shouldn't step four about sockets go after run simulation?
Edit-Adding:

> Edit2 - I guess I should catch all those errors with appropriate report messages to prevent future confusion.

yes, might be useful, or at least disabled the repeated terminal outputs about the simulation being at 0%

I know get these error messages instead (when pressing P)

```
completed
parse_log_file(): Trying to read nodes and elements from file: /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2_BLENDYN.log
parse_log_file(): Found output frequency element.
parse_elements(): Element type output frequency not implemented yet.                 Skipping...
parse_log_file(): Found a structural node.
parse_node(): Parsing node 1
parse_node(): found existing entry in nodes dictionary for node 1. Updating it.
parse_log_file(): Found a structural node.
parse_node(): Parsing node 2
parse_node(): found existing entry in nodes dictionary for node 2. Updating it.
parse_log_file(): Found body element.
parse_body(): Parsing body 1
parse_body(): found existing entry in elements dictionary. Updating it.
parse_log_file(): Found body element.
parse_body(): Parsing body 2
parse_body(): found existing entry in elements dictionary. Updating it.
parse_log_file(): Found revolutepin element.
parse_revolute_pin(): Parsing revolute pin joint 1
parse_revolute_pin(): found existing entry in elements dictionary. Updating it.
parse_log_file(): Found revolutehinge element.
parse_revolute_hinge(): Parsing revolute hinge joint 2
parse_revolute_hinge(): found existing entry in elements dictionary. Updating it.
parse_log_file(): Found struct node dofs element.
parse_elements(): Element type struct node dofs not implemented yet.                 Skipping...
parse_log_file(): Found struct node eqs element.
parse_elements(): Element type struct node eqs not implemented yet.                 Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): Found struct node labels element.
parse_elements(): Element type struct node labels not implemented yet.                 Skipping...
parse_log_file(): row does not contain an element definition. Skipping...
parse_log_file(): row does not contain an element definition. Skipping...

Reached the end of .out file
Could not locate the file /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2_BLENDYN.rfm
Info: MBDyn entities imported successfully

0% completed
```

janga1997 commented:
@louisgag I will remove the line printing simulation status . (Which is what shows 0% continuously.)

If I am not wrong, the console output you posted has no errors. It just shows the nodes being loaded in.
I am clear if the live simulation works, or doesn't work for you?

louisgag commented:
It doesn't, I may be doing something wrong but I'm attempting to follow your instructions as much as possible.. Pressing P seems to (sometimes) update once the 3D view and then nothing, a terminal output saying "Blender Game Engine Started" and a gray 3D view...

janga1997 commented:
@louisgag 
I think I wasted another hour of your vacation. Now that you mentioned a "gray 3D view", I looked over to my steps to realize I never explicitly  mentioned to `Add the nodes to the scene` **after** loading in the log file, **before** starting the sockets.
the same goes for adding the elements to the scene.
If there are no node/element objects in the scene, there's nothing to push all those coordinates to.

I really haven't improved on my oversights.

louisgag commented:
I don't think that's the issue. I can see the nodes. I do 1. select input file 2. run simulation. 3. load log file. 4. add nodes to scene. 5. start socket 6. P

becomes all gray, canceling shows that some keyframes seem to have been imported...

I think it has to do with this error message "Could not locate the file /home/louis/PIF/GSoC/GSoC2017/testingJangaSockets/double_rigid_pendulum_2_BBBB.rfm
Info: MBDyn entities imported successfully"

I have no experience with sockets, could it be that I am suppose to compile MBDyn differently? (ALthough I doubt...)

janga1997 commented:
@louisgag The .rfm file is not the issue either.
Maybe @zanoni-mbdyn could verify.

This is the output you should be seeing **after** you start the sockets.
```
Reading Body(1)
Reading Body(2)
Reading Joint(1)
Reading Joint(2)
Reading StreamOutElem(1)
SocketStreamElem(1): host undefined at line 121, file </home/janga1997/examples/double_rigid_pendulum_2.txt>
SocketStreamElem(1): using default host: 127.0.0.1:9011
starting SocketStreamElem(1)...
Reading Gravity
Gravity: <type> expected at line 126, file </home/janga1997/examples/double_rigid_pendulum_2.txt>; assuming "uniform"

<socket.socket fd=20, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 9011), raddr=('127.0.0.1', 44250)>

Socket Closed
Info: The MBDyn simulation was interrupted

```

The line 
**`<socket.socket fd=20, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 9011), raddr=('127.0.0.1', 44250)>`**

indicates that the sockets launched from python within Blender has been connected to MBDyn.

janga1997 committed:
Remove print messages

zanoni-mbdyn commented:
I was able to see the live animation following your instructions. Good work so far. 

Here's my first comments:

- couldn't we avoid the need of pressing the `Load .log file` the second time by calling the `MBDynReadLog` operator when the `Start Sockets` button is pressed?
- what is the role of the load frequency in this context? Could we set up the communication so that Blendyn is requesting the motion only once every n steps, like we do for the output? This way, by playing with the load frequency, one could see the motion of a (simple) model in real-time or close to it;
- can we make the keyframe insertion optional?

zanoni-mbdyn commented:
Adding to the second comment: it would be also nice to be able to set the output stream from MBDyn to UDP and having a "timer" in Blendyn that tells it to load the content at the right time to see the animation in real-time without having the user fiddling manually with the frequency?

janga1997 commented:
If this PR works for me and @zanoni-mbdyn  (If I remember correctly, we are both on Fedora 25, which has Blender 2.78), then I suppose it must be the Blender versions which are causing the problem for @louisgag .
I will download an tar version of Blender 2,69 and test this on it too.

To address your comments,
1. yes, that's the best way
2. I think we can do that. We still have to send a message to the designated Stream driver to progress the simulation by a step, but we can skip assigning the location/orientation for **n -1** steps.
I am having a little trouble imagining how fractional frequencies would work here, but it seems possible.

3.yes

janga1997 commented:
If this PR works for me and @zanoni-mbdyn  (If I remember correctly, we are both on Fedora 25, which has Blender 2.78), then I suppose it must be the Blender versions which are causing the problem for @louisgag .
I will download an tar version of Blender 2,69 and test this on it too.

To address your comments,
1. yes, that's the best way
2. I think we can do that. We still have to send a message to the designated Stream driver to progress the simulation by a step, but we can skip assigning the location/orientation for **n -1** steps.
I am having a little trouble imagining how fractional frequencies would work here, but it seems possible.

3.yes

janga1997 committed:
Enable optional keyframes

janga1997 committed:
Load frequency in live sim

janga1997 committed:
Merge branch 'master' into live-anim

janga1997 commented:
I think I addressed all your comments.

- The .log file is automatically reloaded on starting the sockets.
- You can speed up/slow down the live simulation by increasing the load frequency, before or even during the simulation.
- Keyframes can be disabled/enabled before, or even during the simulation.

janga1997 committed:
Change load_frequency with keystrokes

janga1997 commented:
I was getting annoyed with manually changing the frequency to speed up/slow down the animation, so I added a commit to change it through keystrokes.
`LEFT_ARROW = Decrease frequency by 1`
`RIGHT_ARROW = Increase frequency by 1`

janga1997 committed:
Parse drivers and output elements

janga1997 commented:
**Note:** Do not try with the latest commit. Its incomplete, and might not work.

janga1997 committed:
support for UNIX sockets

janga1997 commented:
**Update:** Now it works by the filedrivers and output elements from the log file.
The steps are a little simplified now.

1. Select the input file, and `Run Simulation`.
2. Press the `Standard Import` button.
3. Start the sockets with the `Start Sockets` button.

Now, to start the live animation by pressing the **`P`** key once.
To speed up the animation, press **`K`**, and to slow down, press **`J`**. (Arrow keys have other functions in Blender)
To stop the simulation, press **`ESC`**

I've also included support for UNIX sockets.

But now, all the stream elements should be set with **`create: yes`**

I've attached some example files for you( @zanoni-mbdyn ) to try.
[double_rigid_pendulum_2.txt](https://github.com/zanoni-mbdyn/blendyn/files/1185907/double_rigid_pendulum_2.txt) (driver: UNIX, output: INET)
[spinning_top.txt](https://github.com/zanoni-mbdyn/blendyn/files/1187819/spinning_top.txt) (driver: INET, output: INET) This looks the coolest BTW, when you speed up/slow down the animation.
[crank_slider.txt](https://github.com/zanoni-mbdyn/blendyn/files/1187833/crank_slider.txt) (driver:UNIX, output: UNIX)

janga1997 committed:
Minor Fixes

janga1997 committed:
error in render vars

janga1997 committed:
Add Standard import to top too

janga1997 committed:
Multiple drivers

janga1997 committed:
Reassign keys

zanoni-mbdyn commented:
Thank you, I'll test thoroughly in the next few days and give you some final feedback. It definitely seems like the core is there.

@janga1997  can you put the three examples in the `tests/live` folder, with a minimal readme (you could also put it in a comment at the top of the input file)? In doing so, please use the extension `.mbd` or no extension at all, thanks.

janga1997 committed:
add liveanim examples and remove live keyframes

janga1997 committed:
fix key error

janga1997 commented:
@zanoni-mbdyn I removed the option for live keyframes. It wasn't really solving any issue, and you could generate the keyframes through the animate button as before after all.

I added three example files in the test directory for you to test.
- In the `crank_slider.mbd` example, you can increase/decrease the angular velocity of the crank, through the keyboard keys , `LEFT_ARROW`, `RIGHT_ARROW`. 
- You can traverse through the list of drivers in a simulation with `UP_ARROW`, `DOWN_ARROW`.
I've mixed up UNIX sockets and INET sockets in those three files, just to make sure all the combinations work.

zanoni-mbdyn commented:
Great work, thanks.

Just one more little suggestion: can we try to detect the version of MBDyn that is currently selected by the user, and disable the live animation panel in the case it is older than the current one (1.7.2)?

My idea would be to wait to push the live animation feature until the next officially released version of MBDyn, that will support the logging of the stream elements and file drivers, is out. When it is, we'll just change the 1.7.2. to 1.7.3 or whatever the version will be and promote the Live Animation to the master branch. 

Speaking of branches, I'll create (as I had in mind for quite some time now) a devel branch, that will serve as the "bleeding edge" branch for Blendyn: in the future I'd like to follow [this branching model](http://nvie.com/posts/a-successful-git-branching-model/).

janga1997 commented:
@zanoni-mbdyn Did you face any issues when you were traversing through the list of `drivers` with the `UP_ARROW, DOWN_ARROW` keys? Mine seemed to jump a little a bit if pressed too quickly.

And yes, merging these PR's  (this and #32 ) to a devel branch would be a better idea. That should eliminate the need for checking the MBDyn version within Blendyn, right?

zanoni-mbdyn commented:
> @zanoni-mbdyn Did you face any issues when you were traversing through the list of drivers with the UP_ARROW, DOWN_ARROW keys? Mine seemed to jump a little a bit if pressed too quickly.

With the PC on which I am working now, I did not experience lag. I might try again though, because I didn't try to jump very quickly from one driver to another.

> And yes, merging these PR's (this and #32 ) to a devel branch would be a better idea. That should eliminate the need for checking the MBDyn version within Blendyn, right?

Not exactly: I'd like to avoid the situation in which someone tries the live animation with an older version of MBDyn and an unhandled (and very obscure) exception is raised. I think that this check is needed.

zanoni-mbdyn commented:
Oh, and I agree with the removal of the keyframes insertion option.

janga1997 committed:
Check if MBDyn < 1.7.2

janga1997 commented:
@zanoni-mbdyn I placed a check for the MBDyn version. It comes into play after you press the `Run Simulation`.

zanoni-mbdyn commented:
Good, I'll check it in the afternoon.

zanoni-mbdyn commented:
I have noticed that if the Blender session is closed either automatically or for a crash when an MBDyn simulation is running, the mbdyn process is not stopped. This can be a problem, especially using models with `final time: forever`!

janga1997 commented:
@zanoni-mbdyn Noted. Something I have experienced myself quite a few times.
Will put together a commit soon.

janga1997 committed:
stop MBDyn on exit

janga1997 commented:
@zanoni-mbdyn My first commit after a long time !
Getting to the commit, it does stop MBDyn before exit, but there is still the issue of removing the UNIX socket files.
Two ways to go about this
1. Remove UNIX socket files through MBDyn itself, making relevant changes.
2. Agree upon a naming convention with extension `.sock`, for me to remove them before exit.

louisgag commented:
@janga1997 if you go for (2) how are you planning to remove the sockets files?

janga1997 commented:
@louisgag Never mind actually!
 I just realized I already parse the paths of the socket files (regardless of the extension), which is what I use to connect to them.
Since they are already in the **global scene** variable, I can remove them directly on exit.

janga1997 commented:
@louisgag that didn't work out.
Apparently, the **scene** variable is dumped just before exit, so I have no way of accessing the paths of socket files.
This is what is weird with MBDyn right now. 
* If you stop the MBDyn process while it's socket is still in connection (with Blender), MBDyn automatically removes the socket files.
* If the connection is lost before stopping MBDyn, as it happens right now on exit of Blender, the socket file is not removed.

zanoni-mbdyn commented:
@janga1997 Uhm... Interesting... I'll look into the MBDyn code and see if I can pinpoint the issue.

zanoni-mbdyn commented:
I'm closing this for the time being, will resume this when we have some time to work on it...
